Instancer Prototype Roots

andrewkaufman opened the pull request:
This adds control over prototype root locations via a string array, which can be optionally specified using a plug, Constant primvar, or Vertex primvar. Note the default behaviour is backwards compatible, as it uses the prototype scene root children in the absence of a specified root array.

The new `prototypeMode` plug defines how the prototypes map onto each instance. In all cases, a primitive variable will be used to determine which path in the prototypes scene is applied to each instance.

- In "Indexed (Roots List)" mode, the primitive variable must be an integer per-vertex. Optionally, a path in the prototypes scene corresponding to each index can be specified via the `prototypeRootPaths` plug. If no roots are specified, an index of 0 applies the first location			  from the prototypes scene, an index of 1 applies the second, and so on.

- In "Indexed (Roots Variable)" mode, the main primitive variable must be an integer per-vertex, and there must be a separate constant string array primitive variable specifying a path in the prototypes scene corresponding to each index.

- In "Root per Vertex" mode, the primitive variable must be a string per-vertex which will be used to specify a path in the prototypes scene for each instance.

One thing to note, its non-trivial (impossible?) to use `RootPerVertex` mode efficiently, purely in Gaffer... while this will likely be the most commonly used mode at studios using Houdini to generate the input points, I've not a found any way to author an indexed StringVectorData primvar in Gaffer. You can author a non-indexed string primvar using OSLObject, but that generates a unique prototype per instance, which is obviously a bad idea. We could perhaps add a new "index-my-primvar" node to Gaffer, or update `OSLObject` to automatically index string primvars, but both of those felt out-of-scope for this PR.

johnhaddon commented on src/GafferScene/Instancer.cpp:
```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
 
+		void initPrototypes( PrototypeMode &mode, const std::string &index, const std::string &rootsVariable, const StringVectorData *rootsList )
+		{
+			const std::vector<std::string> *rootStrings = nullptr;
```
For this "declare variable and then initialise it differently in different branches" pattern, I've taken to _not_ initialising the variable at the point of declaration. That way the compiler will tell you if you miss an initialisation lower down and then try to use it uninitialised.

andrewkaufman replied:
Addressed in e19c79dbb000bed20de0a249fe3d7ee44d2d0a6d, though to be honest, I'm not sure I like it. I had to add a nonsensical `default` case to satisfy the compiler.

johnhaddon replied:
Interesting. GCC 6 doesn't seem to want the extra `case` - I guess you're using 4.8? But GCC 6 doesn't detect the uninitialised variable if I omit it from one of the case statements either, so obviously I've been trusting the compiler too much. Revert it then I guess...

andrewkaufman replied:
Addressed during the squash rebase

johnhaddon commented on src/GafferScene/Instancer.cpp:
```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
+			for( const auto &root : *rootStrings )
+			{
+				ScenePlug::stringToPath( root, path );
+				roots.emplace_back( new InternedStringVectorData( path ) );
```
Should we be checking the existence of the path, and throwing if it's invalid? I think we probably should?

andrewkaufman replied:
I thought path exists checks were too expensive. Happy to add it though.

johnhaddon replied:
I think here we _should_ be OK, because we don't initialise engines all that frequently. I also have a bit of a plan for improving the performance of existence checks in future so I'm not quite as wary as I once was.

andrewkaufman replied:
Addressed in ea93f16bf034a728ca56fab80ee230bb1c97305d

johnhaddon replied:
I've added one comment on that commit, since I think it has an unnecessary PathScope.

andrewkaufman replied:
Addressed in 09390311f151b02c4ce5972d9119bfd82a490583

johnhaddon commented on src/GafferScene/Instancer.cpp:
```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
+				InternedStringVectorDataPtr name = new InternedStringVectorData;
+				if( path.empty() )
+				{
+					name->writable().emplace_back( std::to_string( i++ ) );
```
There's an `InternedString( i )` constructor that is considerably faster than `InternedString( to_string( i ) )`. Note my comment in the test cases though - I found this replacing of empty paths with numbers really confusing anyway.

andrewkaufman replied:
I imagine we'll be changing this again based on the other comment, but just so I don't lose track of it, addressed in 6bf5ae7f90f7a9e7ee9b5f29772e84c2ae5f4038 for now.

johnhaddon commented on src/GafferScene/Instancer.cpp:
```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
+				inputNames.emplace_back( name );
+			}
+
+			m_names = new Private::ChildNamesMap( inputNames );
```
I hadn't considered that we could use `ChildNamesMap` here - nice!

andrewkaufman replied:
Yeah, it was quite handy, though I'd wished for a constructor that didn't require a vector-of-vectors.

johnhaddon commented on include/GafferScene/Instancer.h:
```diff
@@ -53,15 +53,31 @@ class GAFFERSCENE_API Instancer : public BranchCreator
+
+		Gaffer::StringVectorDataPlug *prototypeRootPathsPlug();
+		const Gaffer::StringVectorDataPlug *prototypeRootPathsPlug() const;
+
```
I find the `prototypeRoots` vs `prototypeRootPaths` distinction a bit hard to keep clear in my head. Any thoughts on alternative namings? Perhaps `prototypesRootsList` and `prototypeRootsVariable`, to tie them directly to the corresponding enum values?

andrewkaufman replied:
In the code I don't mind at all, those new names are fine. But in the UI... switch to IndexedRootsVaraible mode and we have 2 plugs for the user to set, `prototypeIndex` and `prototypeRootsVariable`, both of which are primvar names.. it feels a bit awkward.

johnhaddon replied:
I agree it's a bit awkward to have 2 plugs, but only one suffixed with "Variable". But I think there's something to be said for the explicitness of "RootsList" vs "RootsVariable". Could maybe use a metadata label for the UI? Not sure if by trying to fix one problem I'm just creating another here...

andrewkaufman replied:
Yeah, I really don't like labels masking the true plug names... would it be any better to call them `prototypeRoots` and `prototypeRootsList`? Or is that still too ambiguous?

johnhaddon replied:
I should explain where I found this confusing, in case it helps at all. In the UI it was clear because of the activators. And when reviewing the Instancer internals it was pretty clear too. It was when following the unit tests I found things less clear. When setting up an Instancer programmatically, without the benefit of UI activators, it was hard to keep in my head which plugs mattered at any given time.

> Yeah, I really don't like labels masking the true plug names... would it be any better to call them prototypeRoots and prototypeRootsList? Or is that still too ambiguous?

So we're just renaming `prototypeRootPaths->prototypeRootsList`? That feels like a worthwhile improvement to me, and doesn't seem to introduce any extra awkwardness.

andrewkaufman replied:
Addressed in 63aa83ae508a4c0e4c805a1bbcbc053257801fa5

andrewkaufman commented on python/GafferSceneTest/InstancerTest.py:
```diff
@@ -649,6 +649,355 @@ def testIndices( self ) :
+		script["instancer"]["prototypeRootPaths"].setValue( IECore.StringVectorData( [] ) )
+		self.assertRootsMatchPrototypeSceneChildren( script )
+
+		script["instancer"]["prototypeRootPaths"].setValue( IECore.StringVectorData( [ "", ] ) )
```
Sure, sounds fine. One can always set "/" explicitly to get the other behaviour. To clarify what you're afeter, if the rootpaths are set to `[ "/foo", "", "/bar" ]` are you expecting all points with `index==1` to be skipped? Or were you expecting an implicit index-shift (eg they'd all get "/bar")? I think I favor skipping.

johnhaddon replied:
Skipping sounds good to me.

andrewkaufman replied:
I started implementing this, but I think its going to complicate `Engine::prototypeIndex`. Some points will now have "invalid" indices, so we need to validate them rather than just fetch from `m_indices`. Do you see a simpler way? If not, is the added complexity worth it or should we just handle the trailing "" case (eg `[ "" ]` and `[ "/foo", "/bar", "" ]`)?

johnhaddon replied:
That makes sense, but I think it's worth the extra work to make things bombproof. Seems like it shouldn't be _too_ hard? I would imagine that in `initPrototypes()` we need to initialise a `vector<int> m_prototypeIndexToValidatedPrototypeIndex` (for want of a better name), where we store -1 for invalid indices and adjust the other indices to account for gaps. And then just index into that in `prototypeIndex()`, and skip the invalid indices in the compute for `prototypeChildNamesPlug()`. Did I understand that right?

andrewkaufman replied:
Yeah, it'd be something like that. I'll give it a go then.

Dan and I talked yesterday and thought there might be some way to avoid it if ChildNamesMap had a way to increment the index (ie. track validity itself). But I'm not sure I want to get into all that right now...

johnhaddon replied:
Yeah, it's nice that ChildNamesMap has been helpful, but it is designed for a slightly different use case, so I think we're better off keeping the special case part in the Instancer where it's needed.

andrewkaufman replied:
Addressed in ac8696aad81f52bb6998892fbbaad8e35cc2a5cd

I also added afee26c44ea24be2cc1f11985f4e27ec2ac4f006 to allow the "/" option we described in this discussion. Its unfortunate that it needs a special case in the code, but I think it is useful. I'm not at all attached to using "root" as the name though... Do you have a better suggestion?

andrewkaufman commented on python/GafferSceneUI/InstancerUI.py:
```diff
@@ -95,35 +99,106 @@
+			  be a string per-vertex which will be used to specify a
+			  path in the prototypes scene for each instance.
+
+			  Note -- it is advisable to provide an indexed string
```
Will it be formatted nicely in the tooltip as well?

johnhaddon replied:
Not as nicely as the docs, but serviceably enough.

andrewkaufman replied:
Addressed in f0a3f3fe1043282a94eb5fe978748c705ba5cbc6

andrewkaufman commented on python/GafferSceneUI/InstancerUI.py:
```diff
@@ -95,35 +99,106 @@
-			single object : they can each have arbitrary child
-			hierarchies.
+			Note that the prototypes are not limited to being a single
+			object -- they can have arbitrary child hierarchies.
```
Oh I've no idea what's proper grammatically... I guess I use : for headings only, ; when I'm writing prose and -- when I want to draw someone's attention mid sentence. Here I probably should have replaced the existing : with a ; but I chose -- to be more attention grabbing. I'd 100% defer to @medubelko on what it should be though.

medubelko replied:
A dash (or hyphen, since there are still places were UTF-8 is dangerous) is fine here. This is a reasonable use of it to grab the reader's attention. The rules for typographic conundrums vary widely, and often come down to the editor's personal preference in the style guide. I haven't published a rule for this to give everyone some leeway, but since we haven't been using double hyphens systematically, let's keep it a single hyphen in place of an en dash.

andrewkaufman replied:
Addressed in f0a3f3fe1043282a94eb5fe978748c705ba5cbc6

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -192,9 +191,24 @@ class Instancer::EngineData : public Data
 		size_t prototypeIndex( size_t pointIndex ) const
 		{
-			return m_indices ? (*m_indices)[pointIndex] : 0;
+			return ( m_indices ? (*m_indices)[pointIndex] : pointIndex ) % numPrototypes();
```
> Did you intend to replace : 0 with : pointIndex here? I'm confused because it looks like a break in behaviour compared to before, but I'm not observing the new behaviour I expect to see (cycling pointwise through prototypes if there are no indices).

That's weird, I do see that behaviour here... maybe your test setup wasn't quite right? I think I'd intended that as a fallback; if there's no index primvar then we'll use point order. I do think that's useful (ie. dropping a Seeds on a Plane gives me a round-robin distribution of prototypes). You're right its a change in behaviour and I should have called it out in the PR and the tooltip. I think I lost track of the fact its a change since I did it a month or so ago... I can revert it if you feel strongly about the behaviour change.

Note its not currently throwing for an invalid (as opposed to empty) index primvar name, though it definitely should be.

> I don't know if it has had an impact or not, but moving `% numPrototypes()` here has replaced what was a straight read from a variable with several function calls, for every iteration of a numPoints loop.

That's a bit dramatic. There is only one such loop that's impacted (compute for `prototypeChildNamesPlug`) and several other numPoints loops are not impacted. Still, you're right I haven't profiled and it might well be a regression.

Also, there was never a straight read from a variable. The original code required doing this, which I found incredibly confusing to reason about:

`size_t prototypeIndex = engine->prototypeIndex( i ) % numPrototypes;`

I'll do some profiling to see if I can measure the impact, but I'd argue the new code is considerably clearer; the engine is in charge. Its not much concern since we only have one call-site that needs to do `% numPrototypes`, but at a point before I used `ChildNamesMap` I needed it in multiple places, so it made sense to move it to the engine. Its a const value for each Engine, so I can definitely bake it during `initPrototypes()` to avoid the function call per-point.

johnhaddon replied:
> I think I'd intended that as a fallback; if there's no index primvar then we'll use point order. I do think that's useful (ie. dropping a Seeds on a Plane gives me a round-robin distribution of prototypes). You're right its a change in behaviour and I should have called it out in the PR and the tooltip. I think I lost track of the fact its a change since I did it a month or so ago... I can revert it if you feel strongly about the behaviour change.

I don't have a particularly strong feeling either way, other than about needing to advertise breaking changes. On the one hand, some round-robin variation might be handy. But on the other I suspect the repeating pattern might not be truly useable, and the existence of _some_ indexing might make it a bit harder to determine that the _specific_ indexing you're after hasn't taken hold. On balance I'm inclined to keep things as they were. Any other thoughts @murraystevenson?

> That's a bit dramatic.

I was simply pointing out something I'd noticed had a chance of being related to the performance regression (and `% numPrototypes` _was_ a straight read). As I said, I don't know if it has had an impact or not. I agree, the code is clearer now, so if it's unrelated to performance, then great.

andrewkaufman replied:
> On balance I'm inclined to keep things as they were.

From an off-line discussion, Murray wasn't too fussed either way, and agreed the round-robin approach is of limited use in production scenarios. 
Reverted to original behaviour for missing indices in 291503188b573538156f8b476a975ae59a87b9fb.

> I don't know if it has had an impact or not

I haven't had time to profile yet, but its simple enough to remove, so I've done that in 35fb9852b76609d60354af143479b893c630359a

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
+						{
+							throw IECore::Exception( "prototypeIndex primitive variable has incorrect size" );
+						}
+					}
```
I thought it was clearer to have each case completely handle its data gathering, so worth the small amount of code duplication. Do you disagree? Or were you saying you wish the message of the exception was different so we could identify the issue more easily in production?

johnhaddon replied:
A more descriptive exception would definitely be useful, and the little bit of extra code and double-maintenance involved is what made me wonder about the duplication. I care more about the exception...

andrewkaufman replied:
Addressed in e1195b9ce675209ce7495177f9fafb34d0233784. Note I changed all the primmer lookup exceptions to include the name, not just the one commented on.

johnhaddon replied:
Nice - thanks for going the extra mile.

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
+					const auto *roots = m_primitive->variableData<StringVectorData>( rootsVariable, PrimitiveVariable::Constant );
+					if( !roots )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must be Constant StringVectorData when using IndexedRootsVariable mode" );
```
Including the name makes sense for sure, but I'm not sure its worth distinguishing doesn't exist from wrong type. We've had feedback that its better to give the information up front. Otherwise someone will go create the missing variable only to realize they've made the wrong type or the wrong interpolation. I suppose I could do an additional check after we know we're throwing just to extend the message a bit to say it whether exists or not. Sound ok?

johnhaddon replied:
Sounds great.

andrewkaufman replied:
Addressed in d08e626df406054344b7c49c745f32c52c5479fc

andrewkaufman commented:
> I wonder, have you done any performance profiling at all? With a very quick test using the default prototypes mode, I'm seeing approximately 14% regression against master.

I think some regression is expected; we've increased complexity to gain user flexibility. But 14% feels a bit much... hopefully I can get that down a bit. I don't think I'll have time to profile before Christmas though... do we have a release target date I'd need to get it done by?

> Did you intend to limit this to master only, or would it be useful to target 0.55 instead?

I think the prerequisite #3448 and #3458 are only on master, right?

johnhaddon commented:
> I think some regression is expected; we've increased complexity to gain user flexibility. But 14% feels a bit much... hopefully I can get that down a bit. I don't think I'll have time to profile before Christmas though... do we have a release target date I'd need to get it done by?

No hard date but I'd guess the back end of January. It'll depend a lot on EditScopes progress...

> I think the prerequisite #3448 and #3458 are only on master, right?

So they are - I'd got in my head they were on 0.55 for some reason.

andrewkaufman commented on python/GafferSceneTest/InstancerTest.py:
```diff
@@ -649,6 +649,355 @@ def testIndices( self ) :
+		def updateRoots( roots ) :
+
+			points = script["objectToScene"]["object"].getValue()
+			points["root"] = IECoreScene.PrimitiveVariable( points["root"].interpolation, roots, points["root"].indices )
```
Addressed in 93302f4ed885ac2ee47cb3f4e64e4d1361260c04

andrewkaufman commented on python/GafferSceneUI/InstancerUI.py:
```diff
@@ -95,35 +99,106 @@
+			"description",
+			"""
+			The method used to define how the prototypes map
+			onto each instance. A primitive variable will be
```
Addressed in f0a3f3fe1043282a94eb5fe978748c705ba5cbc6

andrewkaufman commented on python/GafferSceneUI/InstancerUI.py:
```diff
@@ -95,35 +99,106 @@
+			used to determine which path in the prototypes
+			scene is applied to each instance.
+
+			- In "Indexed (Roots List)" mode, the primitive variable
```
Addressed in f0a3f3fe1043282a94eb5fe978748c705ba5cbc6

andrewkaufman commented on python/GafferSceneUI/InstancerUI.py:
```diff
@@ -95,35 +99,106 @@
+			  from the prototypes scene, an index of 1 applies the
+			  second, and so on.
+
+			- In "Indexed (Roots Variable)" mode, the main primitive
```
Addressed in f0a3f3fe1043282a94eb5fe978748c705ba5cbc6

andrewkaufman commented on include/GafferScene/Instancer.h:
```diff
@@ -53,15 +53,31 @@ class GAFFERSCENE_API Instancer : public BranchCreator
 
 		GAFFER_GRAPHCOMPONENT_DECLARE_TYPE( GafferScene::Instancer, InstancerTypeId, BranchCreator );
 
+		enum PrototypeMode
```
Addressed in 73f4761dfbf74230c6a71be096f3ac1006f56dd2

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -537,10 +685,28 @@ void Instancer::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
+				prototypeChildren.emplace_back( name );
+			}
+
+			prototypeRootPaths = new StringVectorData( prototypeChildren );
```
Addressed in 4b4189edbbcbbb48d9446393e545615b029e5cbd

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -915,7 +1080,8 @@ void Instancer::hashBranchChildNames( const ScenePath &parentPath, const ScenePa
 	{
 		// "/instances"
-		h = prototypesPlug()->childNamesHash( ScenePath() );
+		h = IECore::MurmurHash();
```
Addressed in 9d016beb885d44a1ca16a92ac2262fc676662fb5

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -1059,16 +1223,22 @@ void Instancer::prototypeChildNamesHash( const ScenePath &parentPath, const Gaff
 
-Instancer::PrototypeScope::PrototypeScope( const Gaffer::Context *context, const ScenePath &branchPath )
-	:	EditableScope( context )
+Instancer::PrototypeScope::PrototypeScope( const Instancer *instancer, const Gaffer::Context *context, const ScenePath &parentPath, const ScenePath &branchPath )
```
Addressed in 5a21fb6adfde397a44a6bf3690bc9a9541288fed

andrewkaufman replied:
Affects bug fixed in 8a87bc1cf240c9c9b9dbc5185803e20a64e02a83

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -192,9 +191,24 @@ class Instancer::EngineData : public Data
+
+		const ScenePlug::ScenePath &prototypeRoot( const InternedString &name ) const
+		{
+			return runTimeCast<const InternedStringVectorData>( m_roots->members()[m_names->input( name ).index] )->readable();
```
Addressed in fe21fdfffc2cfff723f259091346b3c87902be06

andrewkaufman commented on src/GafferScene/Instancer.cpp:
```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
 			}
 		}
 
+		void initPrototypes( PrototypeMode &mode, const std::string &index, const std::string &rootsVariable, const StringVectorData *rootsList )
```
Addressed in f98aef247cf763ac4dd76e73dbe29d5b088a5299

andrewkaufman commented:
I've addressed all the easy things and left a few open questions before addressing the last couple. Note the first 3 commits remain unchanged, but I had to rebase on master to get Azure going.

It'd be great if you can mark-as-resolved the ones you're satisfied with.

I still haven't done performance testing on my end and won't before the holiday break, but hopefully that won't take too long in the new year.

johnhaddon commented:
> I've addressed all the easy things and left a few open questions before addressing the last couple.
It'd be great if you can mark-as-resolved the ones you're satisfied with.

Great - thanks! I think we've resolved everything except the open questions now, and the one new comment about the PathScope.

> I still haven't done performance testing on my end and won't before the holiday break, but hopefully that won't take too long in the new year.

Will be interesting to see if you find anything we can improve. It seems plausible that we're mainly seeing the overhead of the new accesses to `enginePlug()` per location. If that's the case, it's a pity that we're losing performance for the default case, but there's probably a reasonable argument that in most production cases it's offset against the cost of the wrangling people would have been doing to get their prototypes lined up appropriately at the root of the scene. Anyways, let's see what you can see...

andrewkaufman commented:
You ok if I squash down the commits I have so far, and add new commits to address the outstanding comments?

johnhaddon commented:
> You ok if I squash down the commits I have so far, and add new commits to address the outstanding comments?

Sounds good.

andrewkaufman commented:
Ok, I think I've addressed all comments now (though still no performance testing).

andrewkaufman commented:
I've rebased on master and added 3 new commits. I think only the first commit has been reviewed fully. The next 4 were done prior to Christmas to address review notes, I could squash them down if they look good. The final 3 are new:

- 08e82a8 adds a test showing how attributes are inherited from a prototype root prior to this PR.
- 80e0d7d moves attributes from the prototype roots up one level in the output scene. I've listed this as a breaking change, though I'd expect in most user setups it would go unnoticed, since they are still inherited onto the instances.
- 7ff2075 is an optimization to try and regain some of that performance we've lost by adding more prototype root options.

Here's a summary of my performance testing. I've been using a basic demo setup which seeds points on a plane and instances 3 prototype roots using the default mode (ie. empty roots list). Each prototype has 3 or 4 locations and a single mesh at the leaf. Times are wall time in seconds (as gathered by the `gaffer stats` app). Note I have a 16 core machine.

| Gaffer Version | 10k Instances | 100k  Instances | 1mil Instances |
:-: | :-: | :-: | :-:
0.55.4.0 | 1.033 | 2.171 | 13.970
master | 0.977 | 2.155 | 13.795
after first review (fbb07fc) | 1.026 | 2.447 | 16.315
after moving attributes (80e0d7d) | 1.020 | 2.323 | 15.794
current tip of this branch (7ff2075) | 0.974 | 2.280 | 15.179

johnhaddon commented:
Thankyou sir!
